RecoJets/JetProducers/src/PileupJetIdAlgo.cc

```diff
@@ -139,52 +140,42 @@ void setPtEtaPhi(const reco::Candidate & p, float & pt, float & eta, float &phi
 	phi = p.phi();
 }
 
-// ------------------------------------------------------------------------------------------
 void PileupJetIdAlgo::bookReader()
 {
 	if(etaBinnedWeights_){
           for(int v=0; v<nEtaBins_;v++){
-                std::unique_ptr<TMVA::Reader> etaReader = std::unique_ptr<TMVA::Reader>(new TMVA::Reader("!Color:Silent"));
-                etaReader_.push_back(std::move(etaReader));
-          }
-	} else {
-		reader_ = std::unique_ptr<TMVA::Reader>(new TMVA::Reader("!Color:Silent"));
-	}
-	if(etaBinnedWeights_){
-          for(int v=0; v<nEtaBins_;v++){
+            TMVA::Reader tmpTMVAReader( "!Color:Silent:!Error" );
             for(std::vector<std::string>::iterator it=tmvaEtaVariables_.at(v).begin(); it!=tmvaEtaVariables_.at(v).end(); ++it) {
-                if(  tmvaNames_[*it].empty() ) { 
-			tmvaNames_[*it] = *it;
-		}
-                etaReader_.at(v)->AddVariable( *it, variables_[ tmvaNames_[*it] ].first );
+                if(  tmvaNames_[*it].empty() ) {
+                        tmvaNames_[*it] = *it;
+                }
+                tmpTMVAReader.AddVariable( *it, variables_[ tmvaNames_[*it] ].first );
             }
+            for(std::vector<std::string>::iterator it=tmvaSpectators_.begin(); it!=tmvaSpectators_.end(); ++it) {
+                if(  tmvaNames_[*it].empty() ) {
+                        tmvaNames_[*it] = *it;
+                }
+                tmpTMVAReader.AddSpectator( *it, variables_[ tmvaNames_[*it] ].first );
+            }
+            reco::details::loadTMVAWeights(&tmpTMVAReader,  tmvaMethod_.c_str(), tmvaEtaWeights_.at(v).c_str() );
+            etaReader_.push_back(std::unique_ptr<const GBRForest> ( new GBRForest( dynamic_cast<TMVA::MethodBDT*>( tmpTMVAReader.FindMVA(tmvaMethod_.c_str()) ) ) ) );
```

Comment:
I have to study this syntax before. It is completely new to me and I am not a guru of programming :-( This may take sometime, not sure how much yet.

@lgray: For me to understand, could you comment further on what you mean by "This would reduce the number of times you have to read in the configuration file"? It assumes that there are different MVA configurations or it is for the same MVA configuration for different events?  By "thread" do you mean a call to an MVA configuration? Thanks and sorry for naive questions.